### PR TITLE
BeaconState refactor

### PR DESCRIPTION
**Pull request by dangerousfood**
Refactored the BeaconState to comply with v0.01. Additional changes to validator_registry

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BeaconState.java [hunk 1]
```diff
@@ -13,83 +13,57 @@
...
+  private UnsignedLong current_epoch_start_shard;
+  private UnsignedLong previous_calculation_epoch;
+  private UnsignedLong current_calculation_epoch;
 
   // Finality
-  private long previous_justified_slot;
-  private long justified_slot;
-  private long justification_bitfield;
-  private long finalized_slot;
+  private Bytes32 previous_epoch_seed;
```
Comment (cemozerr) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BeaconState.java:48]: previous_epoch_seed & current_epoch_seed is under //Randomness and committees in the v0.01 spec

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java [hunk 2]
```diff
@@ -442,4 +556,285 @@ public static int clamp(int minval, int maxval, int x) {
...
+    for (int i = 0; i < crosslink_committee.getCommitteeSize(); i++) {
+      int participation_bit = (participation_bitfield[i / 8] >> (7 - (i % 8))) % 2;
+      if (participation_bit == 1) {
+        participants.add(crosslink_committee.getCommittee().get(i));
+      }
+    }
+    return participants;
+  }
+
+  public static Bytes32 generate_seed(BeaconState state, UnsignedLong epoch) {
```
Comment (cemozerr) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java:702]: I think this is not in line with the v0.1 spec: https://github.com/ethereum/eth2.0-specs/blob/v0.1/specs/core/0_beacon-chain.md#generate_seed

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java [hunk 3]
```diff
@@ -442,4 +556,285 @@ public static int clamp(int minval, int maxval, int x) {
...
+    return get_randao_mix(state, epoch.plus(UnsignedLong.valueOf(SEED_LOOKAHEAD)))
+        .and(get_active_index_root(state, epoch));
+  }
+
+  /**
+   * @param validators
+   * @param current_slot
+   * @return The minimum empty validator index.
+   */
+  private int min_empty_validator_index(
```
Comment (cemozerr) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java:712]: I don't think this is in v0.1 either, so we might wanna just take it out.

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BeaconState.java (hunk 1 above)
Comment (dangerousfood) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BeaconState.java:48]: Good catch. Updated

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java (hunk 3 above)
Comment (dangerousfood) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java:712]: I agree but, I will leave that to someone else.

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java (hunk 2 above)
Comment (dangerousfood) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java:702]: Totally agree. I can't fix it now without cascading changes.

**Review by cemozerr — APPROVED** [review]

**Pull request merged**